### PR TITLE
Bookbinder: Check for timing system

### PR DESCRIPTION
**Pull request by algadoreo**
As pointed out in #357, if the timing system isn’t on, the timestamps between wafers don't line up and the system should fall back to making individual books for each wafer. Although the Imprinter will check for a timing system from the database and instruct the Bookbinder, the Bookbinder should still double-check for the presence of timing counters.

* If the counters are missing where they are expected (i.e., the database indicates the timing system was on), an error should be thrown to be caught by Imprinter, which will rerun Bookbinder in single-wafer mode.

* If the counters are present, calculate the correct timestamps and write them into every `.times` field in the output Book. Timestamps are calculated using the logic in load_smurf.py (ver. Jan 23, 2023):

https://github.com/simonsobs/sotodlib/blob/0f3c7b7a0ff882d39cd3e3b8c2258ede0c7ba308/sotodlib/io/load_smurf.py#L2160-L2166

### CODE REVIEW

**Comment by kmharrington** [comment]
@algadoreo  any chance you have some books created with this branch?

**Comment by algadoreo** [comment]
I just made a book in `/mnt/so1/users/jleung/books/16733/obs_1673307576_sat1_1111101` on simons1. It’d be great if you could re-run your tests on it to see if it fixes the issue or not!

**Comment by kmharrington** [comment]
Yeah, looks like I'm getting zero time differences between the different UFMs!

**Comment by mhasself** [comment]
When I look at that book, I see mv9 has two fewer samples than all the other slots.

**Comment by algadoreo** [comment]
There are 2 dropped samples in the original mv9_004.g3 file: t=167330998373897728 (between sample # 1327 and 1328) and t=167331048574397696 (between sample # 101727 and 101728). Otherwise, all other samples line up, so the code is doing what we expect.

I think dropped samples is a whole other problem not within the scope of this PR.

File: sotodlib/io/bookbinder.py [hunk 1]
```diff
@@ -153,19 +210,22 @@ def add(self, f):
...
         if 'tes_biases' in f.keys():
-            if self.bias is None:
-                self.bias = self.create_new_supertimestream(f['tes_biases'])
-            self.bias.times.extend(f['tes_biases'].times)
-            self.bias.data = np.hstack((self.bias.data, f['tes_biases'].data)).astype(np.int32)
+            if self.biases is None:
+                self.biases = so3g.G3SuperTimestream(f['tes_biases'].names, core.G3VectorTime([]),
+                                    np.empty((len(f['tes_biases'].names), 0), dtype=np.int32))
+            self.biases.times.extend(f['tes_biases'].times)
+            self.biases.data = np.hstack((self.biases.data, f['tes_biases'].data)).astype(np.int32)
```
Comment (guanyilun) [inline sotodlib/io/bookbinder.py:223]: I guess the performance fix will go to another PR, but I'm a little confused about the mixed use of int32 and int64 across different fields. Could you clarify that for me?

File: sotodlib/io/bookbinder.py [hunk 2]
```diff
@@ -183,54 +243,35 @@ def rebundle(self, flush_time):
...
 
-            Parameters
-            ----------
-            flush_time : G3Time
-                Output will contain buffered data up to (but not including) this time
-            use_rids : bool
-                Whether the `.names` field should be replaced with readout_ids
+        # Since G3SuperTimestreams cannot be shortened, the data must be copied to
+        # a new instance for rebundling of the buffer
+        signalout = so3g.G3SuperTimestream(self.signal.names, tout, self.signal.data[:,:len(tout)])
```
Comment (guanyilun) [inline sotodlib/io/bookbinder.py:261]: I think it's always good to state our assumptions in terms of assertion statements. For example, are we assuming `len(tout) > 0` here?

File: sotodlib/io/bookbinder.py [hunk 3]
```diff
@@ -417,6 +479,61 @@ def determine_state(self, v_az, v_el, dy=0.001):
 
         self.current_state = state
 
+    def check_times(self, f):
```
Comment (guanyilun) [inline sotodlib/io/bookbinder.py:482]: maybe we should consider a better name for this function because `check_times` makes it sound like a "pure" function but it actually replaces the `.times` field in place in the case that `t[0]>BOOK_START_TIME && t[-1] < BOOK_END_TIME`. Maybe something like `correct_times` is a better name? This in place modification could also be stated clearer in the docstr. Also, I think timestamps from your `get_timestamps` function should be more accurate than the `BOOK_{START,END}_TIME` from imprinter, so I don't think we want to trim frames based on these inaccurate metadata.

**Review by guanyilun — APPROVED** [review]
I like the changes here, so I'm approving with some minor comments for consideration / clarification.

File: sotodlib/io/bookbinder.py (hunk 1 above)
Comment (algadoreo) [reply sotodlib/io/bookbinder.py:223]: Those are the types in the original SMuRF files (int32 for `'data'` and `'tes_biases'` and int64 for `'primary'`). I did ask @mhasself at one point about the discrepancy with the TOD format doc, and then we forgot to update it. Nonetheless, I plan to add type-checking assertions in a subsequent PR where I optimize the array type-casting.

File: sotodlib/io/bookbinder.py (hunk 2 above)
Comment (algadoreo) [reply sotodlib/io/bookbinder.py:261]: I don’t think so (but not 100% certain). G3SuperTimestream can be initialized with 0-length vectors, though I don’t know if it breaks anything down the line (in `FrameProcessor.flush`). Even more reason to have comprehensive tests!

File: sotodlib/io/bookbinder.py (hunk 3 above)
Comment (algadoreo) [reply sotodlib/io/bookbinder.py:482]: I struggled a lot with this one; originally it was `trim_frames`, but then became `adjust_times` when the main purpose shifted to the timestamp calculation. I’m not super happy with `check_times` either, so I’m open to suggestions for a more descriptive name.

You also bring up a good point about the operations being done in-place—do you think it makes more sense to make it a subroutine (no return statements) instead?

File: sotodlib/io/bookbinder.py (hunk 1 above)
Comment (kmharrington) [reply sotodlib/io/bookbinder.py:223]: @jlashner  will know what's actually coming out of the smurf system.

File: sotodlib/io/bookbinder.py [hunk 4]
```diff
@@ -508,13 +619,15 @@ def __call__(self, f):
             List containing frames to be entered into output Book
         """
 
-        def generate_missing_smurf_samples(t, dtype=np.int64):
+        def generate_missing_smurf_samples(f, t):
```
Comment (kmharrington) [inline sotodlib/io/bookbinder.py:622]: If you're working on this function then wouldn't this be the place to fill in missing samples / make the number of samples per UFM all line up? (https://github.com/simonsobs/sotodlib/issues/389) 

It's very inter-related to this PR and is equally going to prevent us from having a working loader function / run analysis off books. So if it's not this PR it needs to be the priority for the next one.

File: sotodlib/io/bookbinder.py (hunk 1 above)
Comment (jlashner) [reply sotodlib/io/bookbinder.py:223]: This is correct, the TES signal and bias data are both 32 bit ints, while the "primary" data which contains timing counters and timestamps are 64 bit.

File: sotodlib/io/bookbinder.py (hunk 4 above)
Comment (algadoreo) [reply sotodlib/io/bookbinder.py:622]: The current gap checker (this function) only checks for gaps only between—not within—frames. I have a good idea of how I can implement the latter, but I think it would be a large enough operation that it should be its own PR.
